components/dashboard/src/service/service.tsx

```diff
@@ -64,4 +57,150 @@ function getGitpodService(): GitpodService {
     return service;
 }
 
-export { getGitpodService };
+let ideFrontendService: IDEFrontendService | undefined;
+export function getIDEFrontendService(service: GitpodService, sessionId: string) {
+    if (!ideFrontendService) {
+        ideFrontendService = new IDEFrontendService(service, sessionId, window.parent);
+    }
+    return ideFrontendService;
+}
+
+export class IDEFrontendService implements IDEFrontendDashboardService.IServer {
+    private workspaceID: string;
+    private instanceID: string | undefined;
+    private user: User | undefined;
+
+    private latestStatus?: IDEFrontendDashboardService.Status;
+
+    private readonly onDidChangeEmitter = new Emitter<IDEFrontendDashboardService.SetStateData>();
+    readonly onSetState = this.onDidChangeEmitter.event;
+
+    constructor(private service: GitpodService, private sessionId: string, private clientWindow: Window) {
+        // we will need hash here
+        const gitpodHostUrl = new GitpodHostUrl(new URL(window.location.toString()));
+        if (!gitpodHostUrl.workspaceId) {
+            throw new Error("no workspace id");
+        }
+        this.workspaceID = gitpodHostUrl.workspaceId;
+        this.processInfo();
+        this.processInstanceUpdate();
+        this.sendSessionID();
+        window.addEventListener("message", (event: MessageEvent) => {
+            if (IDEFrontendDashboardService.isTrackEventData(event.data)) {
+                console.log("=============dashboard isTrackEventData", event.data);
+                this.trackEvent(event.data.msg);
+            }
+            if (IDEFrontendDashboardService.isHeartbeatEventData(event.data)) {
+                console.log("=============dashboard isHeartbeatEventData", event.data);
+                this.activeHeartbeat();
+            }
+            if (IDEFrontendDashboardService.isSetStateEventData(event.data)) {
+                console.log("=============dashboard isSetStateEventData", event.data);
+                this.onDidChangeEmitter.fire(event.data.state);
+            }
+        });
+    }
+
+    private async processInfo() {
+        this.user = await this.service.server.getLoggedInUser();
+        if (this.latestStatus) {
+            this.latestStatus.loggedUserId = this.user.id;
+            this.sendStatusUpdate(this.latestStatus);
+        }
+        const workspace = await this.service.server.getWorkspace(this.workspaceID);
+        this.instanceID = workspace.latestInstance?.id;
+        if (this.instanceID) {
+            this.auth();
+        }
+        // send first status
+        const status = this.getWorkspaceStatus(workspace);
```

Comment:
I think listentostatus already covers it and make sure to order events. I am not sure we should provide any other events, but should rely on it